### ISSUE TITLE
AutoValidateAntiforgeryTokenAuthorizationFilter logs failures with level Information

### ISSUE THREAD

**Issue by clement911** [aspnet/AspNetCore#17478]
Shouldn't the following LogLevel be an Error or at least Warning?

https://github.com/aspnet/AspNetCore/blob/71c91f310f34423e7bff965e6280d74199a0e6fe/src/Mvc/Mvc.ViewFeatures/src/MvcViewFeaturesLoggerExtensions.cs#L74-L77

**Comment by javiercn** [comment]
@clement911 thanks for contacting us.

We don't think this needs to be an error/warning as it is a common situation that can occur and having a higher setting will unnecessarily trigger alerts and pollute the logs.

**Comment by clement911** [comment]
Thanks @javiercn  for the quick response but please allow me to push back a little bit.

There are 2 cases that I can think of why anti forgery validation would fail.

1) The site is under attack.
In that case, I definitely want to know about it so I can assess the threat fully and potentially take action.
If the log level is `Information`, the attack will most likely go un-noticed.

2) The site has a bug and is `POST`ing without a token.
In that case, this is definitely an error.
I understand why a 400 bad request is returned but the developer might not understand what's happening straight away. Logging an error would make it more obvious.

What do you think?

**Comment by javiercn** [comment]
@clement911 A simple common case where your antiforgery token will fail validation.

You open two tabs and render a form at the same time, if the two requests happen in parallel is almost guaranteed that two different tokens get generated and one is wrong as the associated cookie will be overriden in the response.

Another classical scenario is when a user is authenticated, and renders a form and then its authentication credentials expire, at which point the identity of the user is different when it posts the form and will cause an antiforgery failure.

Yet another case is when the user opens one tab, then opens a separate tab and authenticates itself. When it goes back to the original tab and submits the form to the server it will also fail due to the changed identity.

There are many legitimate cases where a request might fail due to an antiforgery issue. We don't want to log errors/warnings in those cases as it will cause unnecessary noise and might trigger alerts or incident response tickets.

Additionally, you don't want to log warnings/errors for scenarios where an attacker can control the input as that might be used to maliciously trigger alarms/alerts and spam the logs.

**Comment by clement911** [comment]
Thanks, I wasn't aware of this scenarios.
I would love to understand this better and ensure we are not making the user experience worse by enabling CRSF validation.

> You open two tabs and render a form at the same time, if the two requests happen in parallel is almost guaranteed that two different tokens get generated and one is wrong as the associated cookie will be overriden in the response.

I wasn't able to reproduce this. Do you expect the CRSF cookie to be reset on every request?
It's not what I am seeing. We get a session cookie valid across all tabs.
The request token is different each time but the cookie token remains the same for the lifetime of the session.

> Another classical scenario is when a user is authenticated, and renders a form and then its authentication credentials expire, at which point the identity of the user is different when it posts the form and will cause an antiforgery failure.
> 
> Yet another case is when the user opens one tab, then opens a separate tab and authenticates itself. When it goes back to the original tab and submits the form to the server it will also fail due to the changed identity.

Do you mean that the CRSF cookie expires?
If you mean that the actual identity cookie expired, then the request would get a 301 to the login page. Am I missing something?

Apologies if this is supposed to be obvious, I am rather to new to CSRF protection.

**Comment by javiercn** [comment]
> I wasn't able to reproduce this. Do you expect the CRSF cookie to be reset on every request?
> It's not what I am seeing. We get a session cookie valid across all tabs.

This only happens when you don't have a cookie and you open two tabs simultaneously. Its not a common issue, if you want to trigger it, try adding a Task.Delay to both responses and opening two tabs in a browser without antiforgery cookie.

> Do you mean that the CRSF cookie expires?
> If you mean that the actual identity cookie expired, then the request would get a 301 to the login page. Am I missing something?

I'm referring to the authentication cookie expiring (for example identity, if you chose remember me) and leave the tab with the form open after you have authenticated long enough to the cookie to expire.

You would only get a 301 in cases where you are issuing a get request if authorization fails, not a post request. You'll also only get a 401/301 if its a form that requires authentication.

> Apologies if this is supposed to be obvious, I am rather to new to CSRF protection.

It's fine, its a complicated topic and we get these questions from time to time.

**Comment by clement911** [comment]
Oh I see now how that could happen thank you.
From a UX perspective this is acceptable as it should be an edge case.

I do have another question though, regarding the identity used to generate the CSRF cookie.
In our app we use several authentication schemes, each with the own identity cookie.
We don't have a default scheme and a user can be authenticated with multiple schemes at the same time.

Which identity will be used to generate the CSRF cookie? Will it use all the identities present at the time when the cookie is generated? If so, does that mean that a new CSRF cookie will be generated any time a user signs in / out with any of the auth scheme?

**Comment by javiercn** [comment]
@clement911 You can check the implementation here https://github.com/aspnet/AspNetCore/blob/master/src/Antiforgery/src/Internal/DefaultClaimUidExtractor.cs#L25

I don't remember the exact details, but we take into account all identities in the principal.

**Comment by clement911** [comment]
After doing some further testing it seems that the cookie is somehow linked to ALL identities present at 
the time that the cookie is generated.

Any time that the user signs in/out with any of our identity provider, the CSRF token becomes invalid.

This is problematic for us because we keep the token in memory in our SPA and use it for a long period of time.

Why are the identity claims somehow linked to the CRSF cookie? I understand the need to generate  a unique cookie, but why should it be validated against the identities?

**Comment by javiercn** [comment]
@blowdart can probably give you a better explanation, but I believe is to prevent some attacks where the attacker could change the cookie+token to force the user to make a valid request or something like that.

I don't remember the exact details, but this has been like this since asp.net

**Comment by blowdart** [comment]
It protects against, mainly, cloud scenarios where you are hosting on a shared domain suffix, where an attacker could generate a CSRF token on another subdomain that would be valid. By linking it to identities which are unique to your application this risk goes away.

Given the increase in subdomain take overs in the cloud these days, it's a growing concern again, and so the behaviour will not change.

**Comment by clement911** [comment]
I see.
On the other hand it's a major problem for us as our users can sign in/out regularly from various providers.

Is it possible to make this configurable in a future release?

Maybe in the options?
`services.AddAntiforgery(options => /*disable identities in cookies*/)`

**Comment by javiercn** [comment]
@clement911 There are a few ways to fix this.

One is to redirect the user after a log-in/log-out to force the regeneration of the token (will happen automatically on a get request if the identity has changed). Given that you are building a SPA application I'm not sure how you are dealing with passing the request token to the browser, but that should work.

A couple more elaborate ways involve:
* Use a hidden iframe to refresh the antiforgery token. 
  * You'll need to make sure this iframe can only be rendered inside your origin.
* Use an API endpoint to retrieve a new set of antiforgery tokens. 
  * You need to make sure this API can't be called trough CORS.

I'm not completely sure of the complete security implications for the 2 and 3rd option, so I would stick to 1 if possible.

**Comment by clement911** [comment]
@javiercn thank you but that sounds quite tricky.

Currently, when serving the SPA we generate a token with `IAntiforgery.GetAndStoreTokens` and then the SPA passes the token on every requests (except GETs).

If that same user opens another tab and authenticates with a different identity, it invalidates the token generated previously.
Given that's it's a SPA we can't redirect the user.
Using an iFrame or an API endpoint seems pretty tricky and as you said it's unclear what it means in terms of security.

We think it would be much more simple and robust (and secure) to generate a CSRF cookie that are not tied to identities in the first place. Don't you agree?
Is there some sort of hook into the CSRF lib that we can inject to remove this tie?

Interestingly enough, we've noticed that adding claims to an existing identity doesn't seem to invalidate the cookie token, but adding a new identity doesn...

**Comment by clement911** [comment]
I opened a new issue to discuss as this issue originally relates to something very different.
https://github.com/aspnet/AspNetCore/issues/17576
